### PR TITLE
Automated next 4 lifecycle tests

### PR DESCRIPTION
**Pull request by shaior**

### CODE REVIEW

**Review by theemj — APPROVED** [review]
looks good to me!

File: tests/lifecycle/tests/lifecycle_shutdown.go [hunk 1]
```diff
@@ -62,11 +63,136 @@ var _ = Describe("lifecycle lifecycle-container-shutdown", func() {
...
+		By("Verify test case status in Junit and Claim reports")
+		err = globalhelper.ValidateIfReportsAreValid(
+			lifeparameters.ShutdownDefaultName,
+			globalparameters.TestCaseFailed)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	// 47385
+	It("Several deployments, several pods, several containers"+
+		"that does not have preStop field configured [negative]", func() {
```
Comment (greyerof) [inline tests/lifecycle/tests/lifecycle_shutdown.go:168]: "that does not have" -> "that don't have"

File: tests/utils/deployment/deployment.go [hunk 2]
```diff
@@ -81,3 +96,42 @@ func RedefineWithPreStopSpec(deployment *v1.Deployment, command []string) *v1.De
 
 	return deployment
 }
+
+func RedefineFirstContainerWithPreStopSpec(deployment *v1.Deployment, command []string) *v1.Deployment {
+	deployment.Spec.Template.Spec.Containers[0].Lifecycle = &corev1.Lifecycle{
```
Comment (greyerof) [inline tests/utils/deployment/deployment.go:101]: You should assert that "len(Containers) > 0" before accessing its first element.

**Review by greyerof — COMMENTED** [review]
Just a couple of personal recommendations/suggestions for this project:
- Helpers files: There's no need to prepend the test suite name as prefix for a helper file like "lifehelper.go"/"nethelper.go" -> it can be just "helper.go". You can have several helpers.go in your code base, as long as they belong to different packages (nethelper, lifecyclehelper...).
- Helper file's function names: There's no need to include the name of the package in the name. For instance, you can have two different "DefineDeployment" functions, defined in different packages:
   nethelper.DefineDeployment()        -> Creates a deployment with network related fields.
   lifecyclehelper.DefineDeployment()  -> Creates a deployment with lifecycle related fields (shutdown, terminationGracePeriod...).

File: tests/lifecycle/tests/lifecycle_shutdown.go [hunk 3]
```diff
@@ -21,12 +20,13 @@ var _ = Describe("lifecycle lifecycle-container-shutdown", func() {
 	})
 
 	// 47311
-	It("One deployment, one pod with one container that has preStop field configured", func() {
+	It("One deployment, one pod with one container that"+
```
[the remark this answers is not in the document]
Comment (shaior) [reply tests/lifecycle/tests/lifecycle_shutdown.go:23]: Done, thank you.

File: tests/lifecycle/lifehelper/lifehelper.go [hunk 4]
```diff
@@ -8,12 +8,26 @@ import (
 )
 
 // DefineLifecycleDeployment Defines basic deployment structure for lifecycle tests.
-func DefineLifecycleDeployment() *v1.Deployment {
-	deploymentStruct := deployment.DefineDeployment(
-		"lifecycleput",
+func DefineLifecycleDeployment(moreThanOneContainer bool, deploymentName string) *v1.Deployment {
```
[the remark this answers is not in the document]
Comment (shaior) [reply tests/lifecycle/lifehelper/lifehelper.go:11]: Done, thank you.

File: tests/lifecycle/tests/lifecycle_shutdown.go [hunk 5]
```diff
@@ -62,11 +63,136 @@ var _ = Describe("lifecycle lifecycle-container-shutdown", func() {
...
+		replicaDefinedDeploymentA := lifehelper.DefineLifecycleDeploymentSeveralPodsWithTwoContainers(
+			"lifecycleputone", 3)
+		preStopDeploymentStructA := deployment.RedefineAllContainersWithPreStopSpec(
+			replicaDefinedDeploymentA, lifeparameters.PreStopCommand)
+
+		err := globalhelper.CreateAndWaitUntilDeploymentIsReady(
+			preStopDeploymentStructA, lifeparameters.WaitingTime)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Define second deployment with preStop field configured", func() {
```
[the remark this answers is not in the document]
Comment (shaior) [reply tests/lifecycle/tests/lifecycle_shutdown.go:114]: it happen because I added func() by mistake to By, thank you!

**Review by greyerof — APPROVED** [review]

File: tests/lifecycle/tests/lifecycle_shutdown.go [hunk 6]
```diff
@@ -62,11 +70,181 @@ var _ = Describe("lifecycle lifecycle-container-shutdown", func() {
...
 		Expect(err).ToNot(HaveOccurred())
 
 	})
 
+	// 47382
+	It("One deployment, several pods, several containers that "+
+		"have preStop field configured", func() {
+
+		By("Define deployment with preStop field configured")
+		replicaDefinedDeployment := globalhelper.AppendContainersToDeployment(
```
Comment (kononovn) [inline tests/lifecycle/tests/lifecycle_shutdown.go:85]: Basically these lines are duplicated across all tests 

```
globalhelper.AppendContainersToDeployment(
			deployment.RedefineWithReplicaNumber(
				deployment.DefineDeployment(
					"lifecycleput",
					lifeparameters.LifecycleNamespace,
					globalhelper.Configuration.General.TnfImage,
					lifeparameters.TestDeploymentLabels), 3),
			2,
			globalhelper.Configuration.General.TnfImage)

		preStopDeploymentStruct := deployment.RedefineAllContainersWithPreStopSpec(
			replicaDefinedDeployment, lifeparameters.PreStopCommand)

```
Can you move them in to a new helper function.
Something like: 
func DefineLifecycleDeploymentPreStop(preStop bool) {
...
}

**Review by kononovn — CHANGES_REQUESTED** [review]
Looks good, just few more steps

**Review by kononovn — APPROVED** [review]
Looks good to me

**Pull request merged**